src/test/components/Viewport.test.js

```diff
@@ -0,0 +1,601 @@
+      });
+    });
+
+    it('cannot scroll past the bottom of the viewport', () => {
+      const { view, getChartProps } = setup();
+      const deltaY = 1000;
+      view.find('.chartViewport').simulate('wheel', getMouseEvent({ deltaY }));
+      view.update();
+      expect(getChartProps().viewport).toMatchObject({
+        viewportLeft: 0,
+        viewportRight: 1,
+        viewportTop: MAX_VIEWPORT_HEIGHT - COMPONENT_HEIGHT,
+        viewportBottom: MAX_VIEWPORT_HEIGHT,
+      });
+    });
+
+    it('cannot scroll past the top of the viewport', () => {
+      const { view, getChartProps } = setup();
+      const deltaY = -100;
+      view.find('.chartViewport').simulate('wheel', getMouseEvent({ deltaY }));
+      view.update();
+      expect(getChartProps().viewport).toMatchObject({
+        viewportLeft: 0,
+        viewportRight: 1,
+        viewportTop: 0,
+        viewportBottom: COMPONENT_HEIGHT,
+      });
+    });
+  });
+
+  describe('mousewheel zooming using shift + mousewheel', function() {
+    // These series of tests are very particular to margin differences, so check
+    // different margin combinations separately.
+    const marginTests = [
+      { marginLeft: 0, marginRight: 0, description: 'no margins' },
+      { marginLeft: 33, marginRight: 0, description: 'with a left margin' },
+      { marginLeft: 0, marginRight: 33, description: 'with a right margin' },
+      { marginLeft: 33, marginRight: 55, description: 'with both margins' },
+    ];
+
+    // Run through the tests.
+    for (const { marginLeft, marginRight, description } of marginTests) {
+      describe(description, function() {
+        // Take into account the margins and the component width to find the inner
+        // component width, which is the active viewport area. All mouse events should
+        // be relative to this space.
+        const innerComponentWidth = COMPONENT_WIDTH - marginLeft - marginRight;
+
+        // |        Viewport        |
+        // |-----------*------------|
+        //             ^
+        //             Mouse position
+        it('zooms the preview selection equally when the mouse is centered', () => {
+          const { scrollAndGetViewport } = setup({
+            marginLeft,
+            marginRight,
+          });
+          const viewport = scrollAndGetViewport({
+            deltaY: -100,
+            shiftKey: true,
+            clientX: LEFT + innerComponentWidth * 0.5 + marginLeft,
+          });
+
+          // Assert that this zooms in equally.
+          expect(viewport.viewportLeft).toBeGreaterThan(0);
+          expect(viewport.viewportRight).toBeLessThan(1);
+          expect(1 - viewport.viewportRight).toBeCloseTo(
+            viewport.viewportLeft,
+            7
+          );
+
+          // Only do an additional viewport top/bottom check here.
+          expect(viewport).toMatchObject({
+            viewportTop: 0,
+            viewportBottom: COMPONENT_HEIGHT + 0,
```

Comment:
why this `+ 0` here ?

Reply:
Just seeing if you're paying attention or not 😂